### PR TITLE
Tool 111 rewrite networking

### PR DESCRIPTION
**Pull request by godrei**
~~Blocked by:
https://github.com/bitrise-steplib/steps-ios-auto-provision/pull/66~~

### CODE REVIEW

File: Gopkg.toml [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+#   name = "github.com/x/y"
+#   version = "2.4.0"
+#
+# [prune]
+#   non-go = false
+#   go-tests = true
+#   unused-packages = true
+
+
+[[constraint]]
```
Comment (BirmacherAkos) [inline Gopkg.toml]: Please remove the example

File: appstoreconnect/appstoreconnect.go [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+
+	log.Debugf("Response:")
+	if err := httputil.PrintResponse(resp); err != nil {
+		log.Debugf("Failed to print response: %s", err)
+	}
+
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
```
Comment (BirmacherAkos) [inline appstoreconnect/appstoreconnect.go]: Please do not forget to handle error

**Comment by BirmacherAkos** [comment]
Do not merge it to the `master` branch, merge it to the `develop` branch.

**Comment by BirmacherAkos** [comment]
Create test cases to make it possible to test.

File: appstoreconnect/appstoreconnect.go [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+
+	c := &Client{
+		keyID:             keyID,
+		issuerID:          issuerID,
+		privateKeyContent: privateKey,
+
+		client:  httpClient,
+		BaseURL: baseURL,
+	}
+	c.common.client = c
```
Comment (lpusok) [inline appstoreconnect/appstoreconnect.go]: Why is this needed?

File: appstoreconnect/appstoreconnect.go [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+	"time"
+
+	"github.com/bitrise-io/bitrise-add-new-project/httputil"
+	"github.com/bitrise-io/go-utils/log"
+	jwt "github.com/dgrijalva/jwt-go"
+	"github.com/google/go-querystring/query"
+)
+
+const (
+	baseURL = "https://api.appstoreconnect.apple.com/"
```
Comment (lpusok) [inline appstoreconnect/appstoreconnect.go]: Make this configurabe from a paramter, to make testing easier.

File: appstoreconnect/appstoreconnect.go [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+)
+
+type service struct {
+	client *Client
```
Comment (lpusok) [inline appstoreconnect/appstoreconnect.go]: Why create a struct for one field?

File: appstoreconnect/appstoreconnect.go [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+	return c, nil
+}
+
+// ensureSignedToken makes sure that the JWT auth token is not expired
+// and return a signed key
+func (c *Client) ensureSignedToken() (string, error) {
+	if c.token != nil {
+		claim := c.token.Claims.(claims)
+		expiration := time.Unix(int64(claim.Expiration), 0)
+		// token is marked valid if it will not expire in the upcoming 10 sec
```
Comment (lpusok) [inline appstoreconnect/appstoreconnect.go]: Can we increase the 10 sec to a more reasonable duration like minutes, to prevent potential failures due to expiration?

File: appstoreconnect/appstoreconnect.go [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+	if v.Kind() == reflect.Ptr && v.IsNil() {
+		return s, nil
+	}
+
+	u, err := url.Parse(s)
+	if err != nil {
+		return s, err
+	}
+
+	qs, err := query.Values(opt)
```
Comment (lpusok) [inline appstoreconnect/appstoreconnect.go]: Check that strings are URL escaped.

File: appstoreconnect/jwt.go [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+	"crypto/ecdsa"
+	"crypto/x509"
+	"encoding/pem"
+	"errors"
+	"time"
+
+	jwt "github.com/dgrijalva/jwt-go"
+)
+
+// signToken signs the JWT token with the give .p8 private key content
```
Comment (lpusok) [inline appstoreconnect/jwt.go]: typo in give (given)

File: appstoreconnect/appstoreconnect.go [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+	"bytes"
+	"encoding/json"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"reflect"
+	"time"
+
+	"github.com/bitrise-io/bitrise-add-new-project/httputil"
```
Comment (lpusok) [inline appstoreconnect/appstoreconnect.go]: IMO not worth to have this dependency for 2 helper functions only.

File: appstoreconnect/appstoreconnect.go (hunk 6 above)
Comment (BirmacherAkos) [reply appstoreconnect/appstoreconnect.go]: Updated to 1 minute

File: appstoreconnect/jwt.go (hunk 8 above)
Comment (BirmacherAkos) [reply appstoreconnect/jwt.go]: Fixed

File: appstoreconnect/appstoreconnect.go (hunk 2 above)
Comment (BirmacherAkos) [reply appstoreconnect/appstoreconnect.go]: Fixed
